C compilation warnings under 3.4 and above

Sadly pip (and tox) hide these  by default so they're hard to notice, but if you do `python setup.py bdist_wheel` you'll see the following compilation warnings:
```
src/zope/index/text/okascore.c:91:15: warning: implicit declaration of function 'PyInt_AsLong' is invalid in C99 [-Wimplicit-function-declaration]
                f = (double)PyInt_AsLong(PyTuple_GET_ITEM(d_and_f, 1));
                            ^
src/zope/index/text/okascore.c:136:6: warning: implicit declaration of function 'Py_InitModule3' is invalid in C99 [-Wimplicit-function-declaration]
        m = Py_InitModule3("okascore", okascore_functions,
            ^
src/zope/index/text/okascore.c:136:4: warning: incompatible integer to pointer conversion assigning to 'PyObject *' (aka 'struct _object *') from 'int' [-Wint-conversion]
        m = Py_InitModule3("okascore", okascore_functions,
          ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Presumably we just need to include a header.

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,6 +5,9 @@ matrix:
         - os: linux
           python: 2.7
         - os: linux
+          python: 2.7
+          env: PURE_PYTHON=1
+        - os: linux
           python: 3.3
         - os: linux
           python: 3.4

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -11,6 +11,10 @@ Changes
   because it was expected that these were usually ``None``.
 - Add support for Python 3.6. See `issue 8
   <https://github.com/zopefoundation/zope.index/issues/8>`_.
+- Make the C implementation of the text index's score function
+  (``zope.text.index.okascore``) importable under Python 3. Previously
+  we would fall back to a pure-Python implementation. See `issue 14
+  <https://github.com/zopefoundation/zope.index/issues/14>`_.
 
 4.2.0 (2016-06-10)
 ------------------

--- a/src/zope/index/text/okapiindex.py
+++ b/src/zope/index/text/okapiindex.py
@@ -189,6 +189,9 @@ to always be 1, and then that quotient is
 regardless of k3's value.  So, in a trivial sense, we are incorporating
 this measure (and optimizing it by not bothering to multiply by 1 <wink>).
 """
+import os
+import platform
+
 from zope.index.text.baseindex import BaseIndex
 from zope.index.text.baseindex import inverse_doc_frequency
 try:
@@ -196,6 +199,15 @@ try:
 except ImportError: #pragma NO COVERAGE
     score = None
 from BTrees.Length import Length
+
+
+_py_impl = getattr(platform, 'python_implementation', lambda: None)
+_is_pypy = _py_impl() == 'PyPy'
+PURE_PYTHON = os.environ.get('PURE_PYTHON') or _is_pypy
+if PURE_PYTHON:
+    score = None
+
+PY2 = str is bytes
 
 class OkapiIndex(BaseIndex):
 
@@ -253,7 +265,7 @@ class OkapiIndex(BaseIndex):
     # D to TF(D,t)*IDF(t) directly, where the product is computed as a float.
     # NOTE:  This may be overridden below, by a function that computes the
     # same thing but with the inner scoring loop in C.
-    if score is None: #pragma NO COVERAGE
+    if score is None:
         def _search_wids(self, wids):
             if not wids:
                 return []
@@ -330,7 +342,8 @@ class OkapiIndex(BaseIndex):
                 d2f = self._wordinfo[t] # map {docid -> f(docid, t)}
                 idf = inverse_doc_frequency(len(d2f), N)  # an unscaled float
                 result = self.family.IF.Bucket()
-                score(result, d2f.items(), docid2len, idf, meandoclen)
+                items = d2f.items() if PY2 else list(d2f.items())
+                score(result, items, docid2len, idf, meandoclen)
                 L.append((result, 1))
             return L
 
@@ -358,4 +371,3 @@ class OkapiIndex(BaseIndex):
         for wid in wids:
             d[wid] = dget(wid, 0) + 1
         return d, len(wids)
-

--- a/src/zope/index/text/okascore.c
+++ b/src/zope/index/text/okascore.c
@@ -35,6 +35,10 @@
 
 #include "Python.h"
 
+#if PY_MAJOR_VERSION >= 3
+#define PY3K
+#endif
+
 #define K1 1.2
 #define B  0.75
 
@@ -128,11 +132,32 @@ static PyMethodDef okascore_functions[] = {
 	{NULL}
 };
 
-void
-initokascore(void)
+#ifdef PY3K
+static struct PyModuleDef moduledef = {
+        PyModuleDef_HEAD_INIT,
+        "okascore",                           /* m_name */
+        "inner scoring loop for Okapi rank",  /* m_doc */
+        -1,                                   /* m_size */
+        okascore_functions,                   /* m_methods */
+        NULL,                                 /* m_reload */
+        NULL,                                 /* m_traverse */
+        NULL,                                 /* m_clear */
+        NULL,                                 /* m_free */
+    };
+
+PyMODINIT_FUNC
+PyInit_okascore(void)
 {
 	PyObject *m;
-
-	m = Py_InitModule3("okascore", okascore_functions,
-			    "inner scoring loop for Okapi rank");
+    m = PyModule_Create(&moduledef);
+    return m;
 }
+#else
+PyMODINIT_FUNC
+initokascore(void)
+{
+	/* XXX: Error checking */
+	Py_InitModule3("okascore", okascore_functions,
+				   "inner scoring loop for Okapi rank");
+}
+#endif

--- a/src/zope/index/text/tests/test_okapiindex.py
+++ b/src/zope/index/text/tests/test_okapiindex.py
@@ -150,8 +150,14 @@ class OkapiIndexTest64(OkapiIndexTestBase, unittest.TestCase):
         import BTrees
         return BTrees.family64
 
+class TestScore(unittest.TestCase):
+
+    def test_score_extension(self):
+        from zope.index.text.okapiindex import PURE_PYTHON, score
+        if PURE_PYTHON:
+            self.assertIsNone(score)
+        else:
+            self.assertIsNotNone(score)
+
 def test_suite():
-    return unittest.TestSuite((
-                      unittest.makeSuite(OkapiIndexTest32),
-                      unittest.makeSuite(OkapiIndexTest64),
-                    ))
+    return unittest.defaultTestLoader.loadTestsFromName(__name__)

